File: Qytetet_Ruby/lib/qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#encoding: utf-8

class Qytetet
    attr_accessor:mazo
    attr_reader:max_jugadores
    attr_reader:num_sorpresas
    attr_accessor:num_jugadores
    attr_accessor:num_casillas
    attr_reader:precio_libertad
    attr_reader:saldo_salida
    attr_accessor:tablero
    attr_accessor:carta_actual
    attr_accessor:jugador_actual
    attr_accessor:dado
    attr_accessor:jugadores
    attr_accessor:indice_jugador_actual
  def initialize
    @mazo = Array.new
    @max_jugadores=4
    @num_sorpresas=10
    @num_casillas=20
    @precio_libertad=200
    @saldo_salida=1000
    @jugadores=Array.new
    
  end
  
  #def actuar_si_en_casilla_edificable
    
  #end
  
  #def actuar_si_en_casilla_no_edificable
  
  #end
  
  #def aplicar_sorpresa
  
  #end
  
  #def cancelar_hipoteca(numero_casilla)
  
  #end
  
  #def comprar_titulo_propiedad
  
  #end
  
  #def edificar_casa(numero_casilla)
  
  #end
  
  #def edificar_hotel(numero_casilla)
  
  #end
  
  #def encarcelar_jugador
  
  #end
  
  def get_carta_actual
    return @carta_actual
  end
  
  def get_dado
    @dado = Dado.new
    return @dado.get_valor
  end
  
  def get_jugador_actual
    return @jugador_actual
  end
  
  def get_jugadores
    return @jugadores
  end
  
  #def get_mazo
  
  #end
  
  #def get_valor_dado
  
  #end
  
  #def hipotecar_propiedad
  
  #end
  
  #def inicializar_cartas_sorpresa
  
  #end
  
  def inicializar_juego(nombres)
    inicializar_jugadores(nombres)
    inicializar_tablero
    inicializar_cartas_sorpresa
  end
  
  private
  
  def inicializar_jugadores(nombres)
    @num_jugadores=nombres.size
    for i in 0..nombres.size()-1
      @jugadores<<Jugador.new(nombres[i])
    end
  end
  
  def inicializar_tablero
    @tablero=Tablero.new
  end
  
  public
  
  #def intentar_salir_carcel(metodo_salir_carcel)
  
  #end
  
  #def jugar
  
  #end
  
  #def mover(num_casilla_destino)
  
  #end
  
  #def obtener_casilla_jugador_actual
  
  #end
  
  def obtener_casillas_tablero
    return @tablero.get_casillas
  end
  
  #def obtener_propiedades_jugador
  
  #end
  
  #def obtener_propiedades_jugador_segun_estado_hipoteca(estado_hipoteca)
  
  #end
  
  #def obtener_ranking
  
  #end
  
  #def obtener_saldo_jugador_actual
  
  #end
  
  #def saldo_jugadores
  
  #end
  
  def set_carta_actual(carta)
    @carta_actual=carta
  end
  
  #def siguiente_jugador
  
  #end
  
  #def tirar_dado
  
  #end
  
  #def vender_propiedad(numero_casilla)
  
  #end
  
  private
  
  def inicializar_cartas_sorpresa
        @mazo<<Sorpresa.new("La seguridad de la celda deja mucho que desear,
             simplemente andas hasta tu libertad",TipoSorpresa::SALIRCARCEL,0)
    
        @mazo<<Sorpresa.new("Te han pillado robando comida en el Lecho de "\
              "Pulgas ,debes ir a las mazmorras de Desembarco del Rey.", TipoSorpresa::IRACASILLA,@tablero.get_carcel)
          
        @mazo<<Sorpresa.new("Viajas al Muro porque quieres mear desde lo alto.",
                TipoSorpresa::IRACASILLA,9)
              
        @mazo<<Sorpresa.new("Viajas a Desembarco del rey, para reunirte con La Mano",TipoSorpresa::IRACASILLA,19)
        
        @mazo<<Sorpresa.new("Los guardias de las mazmorras se han quedado"\
              " dormidos y consigues escapar,quedas libre de las mazmorras.",TipoSorpresa::SALIRCARCEL,0)
          
        @mazo<<Sorpresa.new("Caminas por el bosque y te encuentras a unos bandidos" \
                "y te amenazan con matarte, pierdes 500 monedas de oro.",
                TipoSorpresa::PAGARCOBRAR,-500)
              
        @mazo<<Sorpresa.new("Te encuentras una bolsa con oro, recibes 500",TipoSorpresa::PAGARCOBRAR,500)
        
        @mazo<<Sorpresa.new("Los saqueos han comenzado en las ciudades por escasez de" \
                 "alimentos y los mas pobres asaltan posadas y castillos. Tienes que" \
                 "repararlos, pagas 50 monedas de oro por posada y 100 monedas de oro" \
                 "por castillo.",TipoSorpresa::PORCASAHOTEL,-100)
             
        @mazo<<Sorpresa.new("La gran gestion de La Mano(Tyrion Lannister) hace " \
                 "que las posadas de todo el reino se llenen y los castillos vuelvan a " \
                 "recuperar el honor de otras epocas, recibes 50 monedas de oro por posada" \
                 "y 100 monedas de oro por cada castillo.",TipoSorpresa::PORCASAHOTEL,100)
             
        @mazo<<Sorpresa.new("Reunes a los jugadores de la grandes casas para tratar de " \
                 "disminuir los saqueos en tus tierras, todos los jugadores te dan " \
                "250 monedas de oro en compensacion",TipoSorpresa::PORJUGADOR,250*@num_jugadores)
            
        @mazo<<Sorpresa.new("La guerra con los demas jugadores por tu intento de separar " \
                "tu territorio de los Siete Reinos termina contigo derrotado por ello " \
                "tienes que pagar un tributo por a verla provocado, tienes que pagar " \
                "250 monedas de oro a cada jugadores",TipoSorpresa::PORJUGADOR,-250*@num_jugadores)
  end
  
  public
  
  def consultor_mazo
    for i in 0..@mazo.size()-1
      puts @mazo[i].to_s()
    end
  end
  
end

File: src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;
import java.util.Collections;
/**
 *
 * @author joseng2709
 */
public class Qytetet {
    
    static private Qytetet juego; 
    
    
    static int MAX_JUGADORES=4;
    static int NUM_SORPRESAS=10;
    public static int NUM_CASILLAS=20;
    static int PRECIO_LIBERTAD=200;
    static int SALDO_SALIDA=1000;
    
    private ArrayList<Sorpresa> mazo = new ArrayList<> ();
    private Tablero tablero;
    private Dado dado;
    private Jugador jugadorActual;  // por inicializar
    private ArrayList<Jugador> jugadores; // por inicializar
    private Sorpresa cartaActual;
    private int indiceJugadorActual;
    private EstadoJuego estadoJuego;
    
    
    
    public Qytetet(){
        inicializarTablero();
        jugadores= new ArrayList<> ();
        dado = Dado.getInstance();

    }
    
    public static Qytetet getInstance(){
        if(juego == null){
            juego = new Qytetet();
        }
    return juego;
    }
    
    @Override
    public String toString() {
        return "Qytetet{" + "dado=" + dado + ", tablero=" + tablero + ", jugadorActual=" 
                + jugadorActual + ", jugadores=" + jugadores + ", mazo=" + mazo + ", cartaActual=" 
                + cartaActual + '}';
    }
    
    public Tablero getTablero(){
        return tablero;
    }
    
    void actuarSiEnCasillaEdificable(){}
    
    //void actuarSiEnCasillaNoEdificable(){}
    
    /*public void aplicarSorpresa(){
     boolean tienePropietario = false;
        if(this.cartaActual.getTipoSorpresa() == TipoSorpresa.PAGARCOBRAR){
            this.jugadorActual.modificarSaldo(this.cartaActual.getValor());
        }else if(this.cartaActual.getTipoSorpresa() == TipoSorpresa.IRACASILLA){
            boolean esCarcel = this.tablero.esCasillaCarcel(this.cartaActual.getValor());
            if(esCarcel){
                this.encarcelarJugador();
            }else{
                Casilla nuevaCasilla = this.tablero.obtenerCasillaNumero(this.cartaActual.getValor());
            }
        }else if(this.cartaActual.getTipoSorpresa() == TipoSorpresa.PORCASAHOTEL){
            this.jugadorActual.pagarCobrarPorCasaYHotel(this.cartaActual.getValor());
        }else if(this.cartaActual.getTipoSorpresa() == TipoSorpresa.PORJUGADOR){
            for(int i = 0; i < this.jugadores.size(); i++){
                Jugador jugador = this.jugadores.get(i);
                if(jugador != this.jugadorActual){
                    jugador.modificarSaldo(this.cartaActual.getValor());
                    this.jugadorActual.modificarSaldo(-this.cartaActual.getValor());
                }
            }
        }
        
        if(this.cartaActual.getTipoSorpresa() == TipoSorpresa.SALIRCARCEL){
            this.jugadorActual.setCartaLibertad(cartaActual);
        }else{
            Sorpresa actual = this.cartaActual;
            this.mazo.remove(this.cartaActual);
            this.mazo.add(actual);            
        }
        
        
    return tienePropietario; 
    }*/
    
    //public Boolean cancelarHipoteca(int numeroCasilla){}
    
    public boolean edificarCasa(int numCasilla){
         return jugadorActual.edificarCasa(tablero.obtenerCasillaNumero(numCasilla).getTituloPropiedad());
    }
    
    public boolean edificarHotel(Casilla casilla){
        return jugadorActual.edificarHotel(casilla.getTituloPropiedad());
    }
    
    //private void encarcelarJugador(){}
    
    Sorpresa getCartaAltual(){
        return this.cartaActual;
    }
    
    Dado getDado(){
        return this.dado;
    }
    
    Jugador getJugadorActual(){
        return this.jugadorActual;
    }
    
    public ArrayList<Jugador>getJugadores(){
        return jugadores;
    }
    
    ArrayList getMazo(){
        return mazo;
    }
    
    public int getValorDado(){
        return dado.getValor();
    }
    
    public void hipotecarPropiedad(int numeroCasilla){
        jugadorActual.hipotecarPropiedad(jugadorActual.getPropiedades().get(numeroCasilla));
    }
    
    private void inicializarCartasSorpresa(){
        mazo.add (new Sorpresa ("Te han pillado robando comida en el Lecho de "
                + "Pulgas ,debes ir a las mazmorras de Desembarco del Rey.",tablero.getCasillaCarcel().getNumeroCasilla(),
                TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Viajas al Muro porque quieres mear desde lo alto.",9,
                TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa ("Viajas a Desembarco del rey, para reunirte con La"
                + "Mano",19,TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa ("Los guardias de las mazmorras se han quedado "
                + " dormidos y consigues escapar,quedas libre de las mazmorras.",0,
                TipoSorpresa.SALIRCARCEL));
        mazo.add(new Sorpresa ("Caminas por el bosque y te encuentras a unos bandidos"
                + "y te amenazan con matarte, pierdes 500 monedas de oro.",-500,
                TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa ("Te encuentras una bolsa con oro, recibes 500",
                500,TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa ("Los saqueos han comenzado en las ciudades por escasez de"
                + "alimentos y los más pobres asaltan posadas y castillos. Tienes que"
                + "repararlos, pagas 50 monedas de oro por posada y 100 monedas de oro"
                + "por castillo.",-50*1-100*1,TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("La gran gestion de La Mano(Tyrion Lannister) hace "
                + "que las posadas de todo el reino se llenen y los castillos vuelvan a "
                + "recuperar el honor de otras epocas, recibes 50 monedas de oro por posada"
                + "y 100 monedas de oro por cada castillo.",50*1+100*1,TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("Reunes a los Señores de la grandes casas para tratar de "
                + "disminuir los saqueos en tus tierras, todos los Señores te dan "
                + "250 monedas de oro en compensación",250*1,TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("La guerra con los demas Señores por tu intento de seperar "
                + "tu territorio de los Siete Reinos termina contigo derrotado por ello "
                + "tienes que pagar un tributo por a verla provocado, tienes que pagar "
                + "250 monedas de oro a cada jugador",-250*1,TipoSorpresa.PORJUGADOR));
    } 

    public void iniciarJuego(ArrayList<String> nombres){
        this.inicializarCartasSorpresa();
        this.inicializarTablero();
        this.inicializarJugadores(nombres);
        this.salidaJugadores(); 
    }
    
    private void inicializarJugadores(ArrayList<String> nombres){
        
        for(int i = 0; i < nombres.size(); i++){
            this.jugadores.add( new Jugador(nombres.get(i)) );
        }
}
 
    private void inicializarTablero(){
        tablero = new Tablero();
    }
    
    //private Boolean intentarSalircarcel(){}
    
    private void jugar(){
        int resultado=tirarDado();
        Casilla casillaDestino=tablero.obtenerCasillaFinal(jugadorActual.getCasillaActual(), resultado);
        mover(casillaDestino.getNumeroCasilla());
    }
    
    void mover(int numCasillaDestino){
        jugadorActual.setCasillaActual(tablero.obtenerCasillaNumero(numCasillaDestino));
    }
    
    public Casilla obtenerCasillaJugadorActual(){
        return jugadorActual.getCasillaActual();
    }
    
    public ArrayList<Casilla> obtenerCasillasTablero(){
        return tablero.getPropiedades();
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugador(){
        ArrayList<Integer> devolver;
        devolver=new ArrayList<> ();
        
        for(int i=0;i<(this.jugadorActual.getPropiedades().size()-1);i++){
            for(int j=0;j<19;j++){
                if(this.jugadorActual.getPropiedades().get(i)==tablero.obtenerCasillaNumero(j).getTituloPropiedad())
                    devolver.add(j);
            }
        }
        
        return devolver;
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugadorSegunEstadoHipoteca(Boolean estadoHipoteca){
        ArrayList<Integer> devolver;
        devolver=new ArrayList<> ();
        
        for(int i=0;i<(this.jugadorActual.getPropiedades().size()-1);i++){
            for(int j=0;j<19;j++){
                if(this.jugadorActual.obtenerPropiedades(estadoHipoteca).get(i)==tablero.obtenerCasillaNumero(j).getTituloPropiedad())
                    devolver.add(j);

            }
        }
        return devolver;
    }
    
    public Boolean jugadorActualEnCalleLibre(){
        Boolean devolver;
        if(jugadorActual.estoyEnCalleLibre()==true)
            devolver=true;
        else
            devolver=false;
        
        return devolver;
    }
    
    public Boolean jugadorActualEncarcelado(){
        return jugadorActual.getEncarcelado();
    }
       
    public void obtenerRanking(){ 
        Collections.sort(this.jugadores);
    }
    
    public int obtenerSaldoJugadorActual(){
        return jugadorActual.getSaldo();
    }
    
    private void salidaJugadores(){
        for(int i = 0; i < this.jugadores.size(); i++){
            jugadores.get(i).setCasillaActual(tablero.obtenerCasillaNumero(0));
            jugadores.get(i).setSaldo(7500);            
        }
    }   
    
    private void setCartaActual(Sorpresa cartaActual){
        this.cartaActual=cartaActual;
    }
    
    public void setEstadoJuego(EstadoJuego estadoJuego){
        this.estadoJuego=estadoJuego;
    }
    
    public void siguienteJugador(){
        int posicion=0;
        Boolean encontrado=false;
        
        for(int i=0;i<this.jugadores.size() && !encontrado;i++){
            if(jugadorActual == jugadores.get(i)){
                posicion = i; 
                encontrado = true;
            }
        }
        if(posicion==this.jugadores.size())
            jugadorActual=this.jugadores.get(0);
        else
            jugadorActual=this.jugadores.get(posicion+1);
        
        if(jugadorActual.getEncarcelado()==true)
            setEstadoJuego(EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD);
        else
            setEstadoJuego(EstadoJuego.JA_PREPARADO);
        
    }
    
    int tirarDado(){
       return dado.Tirar();
    }
    
    public Boolean venderPropiedad(int numeroCasilla){
        return jugadorActual.venderPropiedad(tablero.obtenerCasillaNumero(numeroCasilla));
    }
        
    
}
